src/Language/PureScript/TypeChecker.hs

```diff
@@ -155,18 +155,28 @@ checkDuplicateTypeArguments args = for_ firstDup $ \dup ->
   firstDup = listToMaybe $ args \\ nub args
 
 checkTypeClassInstance
-  :: (MonadState CheckState m, MonadError MultipleErrors m, MonadWriter MultipleErrors m)
-  => ModuleName
+  :: (MonadState CheckState m, MonadError MultipleErrors m)
+  => TypeClassData
+  -> Int -- type argument index
   -> Type
   -> m ()
-checkTypeClassInstance _ (TypeVar _) = return ()
-checkTypeClassInstance _ (TypeLevelString _) = return ()
-checkTypeClassInstance _ (TypeConstructor ctor) = do
-  env <- getEnv
-  when (ctor `M.member` typeSynonyms env) . throwError . errorMessage $ TypeSynonymInstance
-  return ()
-checkTypeClassInstance m (TypeApp t1 t2) = checkTypeClassInstance m t1 >> checkTypeClassInstance m t2
-checkTypeClassInstance _ ty = throwError . errorMessage $ InvalidInstanceHead ty
+checkTypeClassInstance cls i = check where
+  -- is this argument fully determined via fundeps
+  isFunDepDetermined = (Any False, Any True) == foldMap determining (typeClassDependencies cls)
+    where determining fd = (Any (i `elem` fdDeterminers fd),
```

Comment:
You're right, with the approximation I wrote down, they wouldn't behave the same. In terms of a graph, I would say the correct general version is:

    There is some node (let's call it an initial node, for want of a better name, but it'll be the one which the solver cases on to make its decision) which is not determined by any fundep, and a path from that node to the one we're interested in.

In the case of fundeps with multiple inputs, it's slightly trickier:

    Working backwards from the target, every path ends in an initial node.

Either way, this means that every pair of instances with different types in the target node will be forced to differ in some initial node, or else will be overlapping.